django/db/models/expressions.py

```diff
@@ -665,6 +665,8 @@ def copy(self):
 
 class Value(Expression):
     """Represent a wrapped value as a node within an expression."""
+    for_save = False
```

Comment:
Do we have related tests changes?

Reply:
I think there was, I'll investigate it later today.

Reply:
@felixxm a lot of tests involving `Value` for non-saving purposes crash with an `AttributeError` if not specified here since `for_save` is only assigned in `resolve_expression` and we've got usages of non-resolved `Value` expression in the ORM.

The crash occurs here

https://github.com/django/django/blob/92a2521f786821318564045377f50571dd918082/django/db/models/expressions.py#L718-L719

which is almost always reached now that `output_field` is resolves for all stdlib types.

Reply:
Actually I just pushed a commit that replaces all usages of unresolved `Value` during SQL compilation and replace the commit that added `for_save=True`.

Reply:
It seems odd to remove `output_field` and then have to replace a good many of those with manual resolution of expressions with `.resolve_expression()` all over the place. If a default value of `for_save` can avoid that and is acceptable, is that not better?

I feel we should also document that it is no longer necessary to specify `output_field` for `Value` in many cases, but - if we are keeping this approach - that `.resolve_expression()` ought to be called in `.as_sql()` methods instead.

After all, this feels like we are trying to make things easier to use and less surprising. We should trumpet that.

Reply:
I'm open to whatever approach we prefer here, trying to compile unresolved expression is definitely an anti-pattern but it's likely to be used in the wild as https://github.com/django/django/pull/11359/commits/7c687ac5e4546ec044206a8ff96d2347d96506a1 demonstrates. I agree that maintaining the ability to compile unresolved `Value` is likely the best approach. If we decide to _correct_ that it should go through a deprecation cycle.

@pope1ni one thing to keep in mind though is that unresolved `Value` instances that provided an `output_field` [always crashes on compilation attempts](https://github.com/django/django/blob/92a2521f786821318564045377f50571dd918082/django/db/models/expressions.py#L718-L719). The ones we had in the code base only happened to pass because we didn't specify an explicit `output_field`. That's an unexpected behaviour to me.

> I feel we should also document, in release notes, that it is no longer necessary to specify output_field for Value in many cases ...

This is what we plan to do https://github.com/django/django/pull/11359#issuecomment-640560780

@felixxm any objection to restore the commit that added `for_save = False` instead of 7c687ac5e4546ec044206a8ff96d2347d96506a1? I guess I could also add a comment mentioning _why_ it's defined and add a regression test for now and do a follow up PR that deprecates doing so.

Reply:
> @felixxm any objection to restore the commit that added `for_save = False` instead of [7c687ac](https://github.com/django/django/commit/7c687ac5e4546ec044206a8ff96d2347d96506a1)?

Not really, we can restore it.